### ISSUE TITLE
missing features?

### ISSUE THREAD

**Issue by kunalkathuria** [broadinstitute/CellBender#244]
Hello!

Thank you for creating Cellbender. I had a question about the filtered h5 that I am reading through the data-loading workaround in scanpy. It seems that the anndata object only contains a subset of the raw input genes (e.g. 3000 of 36000 total). This didn't happen for one of my datasets but is happening consistently in this one. Could you please help? Thanks!

Kunal

**Comment by sjfleming** [comment]
Hi Kunal, yes happy to try to figure out what's going on there.

Can you post the cellbender log file here (or at least the first part up to where it starts reporting the training ELBO)?

And can you post what commands you use for dataloading and then what `adata` looks like?

**Comment by kunalkathuria** [comment]
Absolutely!

cellbender:remove-background: Command:
cellbender remove-background --input ./ --output cellbender_output.h5 --expected-cells 5000 --total-droplets-included 732023 --fpr 0.01 --epochs 150 --cuda --learning-rate .00001
cellbender:remove-background: 2023-08-08 22:24:07
cellbender:remove-background: Running remove-background
cellbender:remove-background: Loading data from directory ./
cellbender:remove-background: CellRanger v3 format
cellbender:remove-background: Trimming dataset for inference.
cellbender:remove-background: Including 32905 genes that have nonzero counts.
cellbender:remove-background: Prior on counts in empty droplets is 59
cellbender:remove-background: Prior on counts for cells is 9506
cellbender:remove-background: Excluding barcodes with counts below 29
cellbender:remove-background: Using 5000 probable cell barcodes, plus an additional 61076 barcodes, and 0 empty droplets.
cellbender:remove-background: Largest surely-empty droplet has 29 UMI counts.
cellbender:remove-background: Running inference...
cellbender:remove-background: [epoch 001]  average training loss: 3574.2837
cellbender:remove-background: [epoch 002]  average training loss: 3127.6545  (19.6 seconds per epoch)
...
cellbender:remove-background: [epoch 150]  average training loss: 1855.4312
cellbender:remove-background: [epoch 150] average test loss: 1863.5292
cellbender:remove-background: Inference procedure complete.
cellbender:remove-background: Training succeeded
cellbender:remove-background: 2023-08-08 23:13:54
cellbender:remove-background: Preparing to write outputs to file...
cellbender:remove-background: Optimal posterior regularization factor = 1.67
cellbender:remove-background: Succeeded in writing CellRanger v3 format output to file cellbender_output.h5
cellbender:remove-background: Succeeded in writing CellRanger v3 format output to file cellbender_output_filtered.h5
cellbender:remove-background: Saved cell barcodes in cellbender_output_cell_barcodes.csv
cellbender:remove-background: Saved summary plots as cellbender_output.pdf
cellbender:remove-background: Completed remove-background.

---
import dataLoadingScanpy_workaround as dl
adata_T = dl.anndata_from_h5(file=path + sampleID + "/cellbender_output_filtered.h5") 
---
AnnData object with n_obs × n_vars = 11448 × 14153
    obs: 'ambient_expression', 'feature_type', 'gene_id', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'n_counts', 'n_genes', 'sample name', 'sample type'
    var: 'latent_RT_efficiency', 'latent_cell_probability', 'latent_scale', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts'
    uns: 'contamination_fraction_params', 'fraction_data_used_for_testing', 'lambda_multiplier', 'log1p', 'overdispersion_mean_and_scale', 'target_false_positive_rate', 'test_elbo', 'test_epoch', 'training_elbo_per_epoch'
    varm: 'latent_gene_encoding'

Thanks!

Kunal

**Comment by sjfleming** [comment]
Hi Kunal, I have to admit, I am a bit confused by this...

So there was at one time a bug where I had limited the feature index to be an unsigned 16-bit integer.  The maximum possible number of features was then 65,535.  If there are more total features than that in the raw data, then this could be the problem (if you used a version of cellbender before this change)
https://github.com/broadinstitute/CellBender/blob/d92cfc5a55c8b1771348468035993c52df975170/cellbender/remove_background/infer.py#L417

Which version of cellbender are you running?

... hmmm but it looks like this was already fixed as of version 0.2.0

Any chance you could send me that h5 file?

Can you print the output of
```python
import tables

with tables.open_file('cellbender_output.h5') as f:
    print(f)
```
?

**Comment by kunalkathuria** [comment]
Hi Stephen,

Yes, only 36601 features. Attaching the outputs here. Appreciate your help.

[table_output_filtered.txt](https://github.com/broadinstitute/CellBender/files/12401697/table_output_filtered.txt)
[table_output_regular.txt](https://github.com/broadinstitute/CellBender/files/12401698/table_output_regular.txt)


Best,

Kunal

**Comment by sjfleming** [comment]
Hi @kunalkathuria , okay so this seems to be a problem with the data loading function (or something that's been done after loading the data).

In the raw h5 file, I see that there are in fact 36601 features.
<img width="878" alt="image" src="https://github.com/broadinstitute/CellBender/assets/10214815/5e6bb88f-47c6-41e6-9036-f182c21413b5">

Can you try to run this data loading function?  Maybe it's possible there was a bug in the old one:
https://github.com/broadinstitute/CellBender/blob/4990df713f296256577c92cab3314daeeca0f3d7/cellbender/remove_background/downstream.py#L29

```python
from cellbender.remove_background.downstream import anndata_from_h5

adata = anndata_from_h5('output.h5')
adata
```

**Comment by kunalkathuria** [comment]
Ah, it works now, thank you so much! Not sure if the old script was buggy, as I made some changes at my end too while implementing above. Appreciate your help.

Kunal
